crates/networking/p2p/sync_supervisor.rs

```diff
@@ -0,0 +1,163 @@
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, RwLock,
+};
+
+use ethrex_blockchain::Blockchain;
+use ethrex_common::H256;
+use ethrex_storage::{error::StoreError, Store};
+use tokio::sync::Mutex;
+use tokio_util::sync::CancellationToken;
+
+use crate::{
+    kademlia::KademliaTable,
+    sync::{SyncManager, SyncMode},
+};
+
+pub enum SyncStatus {
+    Active(SyncMode),
+    Inactive,
+}
+
+/// Abstraction to interact with the active sync process without disturbing it
+#[derive(Debug)]
+pub struct SyncSupervisor {
+    /// This is also held by the SyncManager and allows tracking it's latest syncmode
+    /// It is a READ_ONLY value, as modifications will disrupt the current active sync progress
+    snap_enabled: Arc<AtomicBool>,
+    syncer: Arc<Mutex<SyncManager>>,
+    last_fcu_head: RwLock<H256>,
+    store: Store,
+}
+
+impl SyncSupervisor {
+    pub fn new(
+        peer_table: Arc<Mutex<KademliaTable>>,
+        sync_mode: SyncMode,
+        cancel_token: CancellationToken,
+        blockchain: Arc<Blockchain>,
+        store: Store,
+    ) -> Self {
+        let snap_enabled = Arc::new(AtomicBool::new(matches!(sync_mode, SyncMode::Snap)));
+        let syncer = Arc::new(Mutex::new(SyncManager::new(
+            peer_table,
+            snap_enabled.clone(),
+            cancel_token,
+            blockchain,
+        )));
+        Self {
+            snap_enabled,
+            syncer,
+            last_fcu_head: RwLock::new(H256::zero()),
+            store,
+        }
+    }
+
+    /// Creates a dummy SyncSupervisor for tests where syncing is not needed
+    /// This should only be used in tests as it won't be able to connect to the p2p network
+    pub fn dummy() -> Self {
+        Self {
+            snap_enabled: Arc::new(AtomicBool::new(false)),
+            syncer: Arc::new(Mutex::new(SyncManager::dummy())),
+            last_fcu_head: RwLock::new(H256::zero()),
+            store: Store::new("temp.db", ethrex_storage::EngineType::InMemory)
+                .expect("Failed to create test DB"),
+        }
+    }
+
+    /// Updates the last fcu head. This may be used on the next sync cycle if needed
+    pub fn set_head(&self, fcu_head: H256) {
+        if let Ok(mut last_fcu_head) = self.last_fcu_head.write() {
+            *last_fcu_head = fcu_head
+        }
+    }
+
+    /// Returns the current sync status, either active or inactive and what the current syncmode is in the case of active
+    /// It will also start the next cycle if there is a pending sync
+    pub fn status(&self) -> Result<SyncStatus, StoreError> {
+        // Check current sync status and act accordingly
+        Ok(match self.sync_status_internal()? {
+            SyncStatusInternal::Inactive => SyncStatus::Inactive,
+            SyncStatusInternal::Active => SyncStatus::Active(self.sync_mode()),
+            SyncStatusInternal::Pending => {
```

Comment:
it is a bit weird that a read method starts an action. Shouldn't this code be inside `set_head ` (when setting head, if pending then start cycle)?

Reply:
I figured `status` is called more often, ideally `start_sync` should loop through sync cycles if needed to avoid having downtime, I think we can do that, ill give it a try

Reply:
Update: implemented in ([72fdbfd](https://github.com/lambdaclass/ethrex/pull/2303/commits/72fdbfd36e8e69a0a03d2b4ddda67029c9c77d80)

Reply:
That behaviour has been removed in [72fdbfd](https://github.com/lambdaclass/ethrex/pull/2303/commits/72fdbfd36e8e69a0a03d2b4ddda67029c9c77d80)

Reply:
Ah cool, it seems I was looking at an outdated diff